### PR TITLE
Rd 9298 angels lambda node tracer trace mongodb 4 x

### PR DESCRIPTION
**Pull request by therightstuff**
update the node tracer for compatibility with mongodb client version 4.x

### CODE REVIEW

File: src/hooks/mongodb4x.test.js [hunk 1]
```diff
@@ -0,0 +1,86 @@
+import { HandlerInputsBuilder } from '../../testUtils/handlerInputsBuilder';
+import { getMockedMongoClient, wrapMongoCollection } from '../../testUtils/mongo4xMocker';
```
Comment (mmanciop) [inline src/hooks/mongodb4x.test.js:2]: I positively _hate_ mocks in instrumentation tests. They hide infinite bugs and regressions. Can't we rather run the tests against the actual package and maybe, just maybe, a sprinkling of https://github.com/testcontainers/testcontainers-node?

File: src/hooks/mongodb4x.js [hunk 2]
```diff
@@ -0,0 +1,58 @@
+import * as logger from '../logger';
```
Comment (doriaviram) [inline src/hooks/mongodb4x.js:1]: All new files should be typescript with types

File: src/hooks/mongodb.js [hunk 3]
```diff
@@ -1,98 +1,61 @@
...
 
-  return {
-    addPendingRequest,
-    popPendingRequestSpanId,
-  };
-})();
+  /* istanbul ignore next : there's no point in testing this */
+  if (!mongoClients) {
+    logger.debug('MongoDB clients not found');
+    return;
```
Comment (doriaviram) [inline src/hooks/mongodb.js:16]: You test setting the expected behavior. There is a point to check there is no fail. 
And you don't want to be the first one who adding `istanbul ignore` comments :)

File: src/hooks/mongodb.js [hunk 4]
```diff
@@ -1,98 +1,61 @@
...
-      mongoRequestId: requestId,
-      mongoOperationId: operationId,
-      mongoConnectionId: connectionId,
+  mongoClients.forEach((mongoClientLibrary) => {
+    let mongoClientVersion = 'unknown';
+    if (mongoClientLibrary.instrument) {
+      mongoClientVersion = '3.x';
+    } else if (mongoClientLibrary.MongoClient.prototype.on) {
+      mongoClientVersion = '4.x';
+    }
```
Comment (doriaviram) [inline src/hooks/mongodb.js:25]: Try to clean up the code; instead of doing this if only for a switch case, you can do then in once
```
if (mongoClientLibrary.instrument) {
    instrumentMongo3x()
}
if (mongoClientLibrary.MongoClient.prototype.on) {
     instrumentMongo4x()
}
```

File: src/hooks/mongodb3x.test.js [hunk 5]
```diff
@@ -1,31 +1,33 @@
...
-} from '../../testUtils/mongoMocker';
+import { getMockedMongoClient, wrapMongoCollection } from '../../testUtils/mongo3xMocker';
 import { MongoSpanBuilder } from '../../testUtils/mongoSpanBuilder';
 import { SpansContainer, TracerGlobals } from '../globals';
 import { hookMongoDb } from './mongodb';
 
 const DUMMY_URL = 'mongodb://localhost:27017/myproject';
 
 describe('mongodb', () => {
+  let connection;
```
Comment (doriaviram) [inline src/hooks/mongodb3x.test.js:10]: Why extract this to global, it's going to create flakiness when running parallel

File: src/hooks/mongodb3x.test.js [hunk 6]
```diff
@@ -1,31 +1,33 @@
...
-    hookMongoDb(mongoLib);
-    const connection = await promisifyMongoFunc(mongoClient.connect)(DUMMY_URL, {});
+    hookMongoDb(mongoClientLibrary);
+    connection = await mongoClient.connect(DUMMY_URL, {});
     const collection = connection.db().collection('documents');
     wrapMongoCollection(collection, 'insert');
 
     const docs = [{ a: 1 }, { a: 2 }, { a: 3 }];
-    await promisifyMongoFunc(collection.insert)(docs);
+    await collection.insert(docs);
```
Comment (doriaviram) [inline src/hooks/mongodb3x.test.js:30]: Why did you add the `promisifyMongoFunc` to the mock? the actual library is not promised as far as I know

File: src/hooks/mongodb3x.test.js [hunk 7]
```diff
@@ -1,31 +1,33 @@
 import { HandlerInputsBuilder } from '../../testUtils/handlerInputsBuilder';
```
Comment (doriaviram) [inline src/hooks/mongodb3x.test.js:1]: Generally comment, this file should not be changed. Changing the tests make us work hard to check if the old logic is broken

File: src/hooks/mongodb4x.test.js [hunk 8]
```diff
@@ -0,0 +1,109 @@
...
+  [[DUMMY_URL]],
+  [[DUMMY_URL, {}]],
+  [[DUMMY_URL, { url: DUMMY_URL }]],
+  [[DUMMY_URL, undefined]],
+];
+
+const invalidSimpleFlowArguments = [[[0, 1]], [[0, 'abc']], [[0, () => {}]]];
+
+describe('mongodb', () => {
+  let connection;
```
Comment (doriaviram) [inline src/hooks/mongodb4x.test.js:19]: Same here, avoid global variables, to allow parallel run

File: src/hooks/mongodb4x.test.js [hunk 9]
```diff
@@ -0,0 +1,109 @@
...
+const validSimpleFlowArguments = [
+  [[DUMMY_URL]],
+  [[DUMMY_URL, {}]],
+  [[DUMMY_URL, { url: DUMMY_URL }]],
+  [[DUMMY_URL, undefined]],
+];
+
+const invalidSimpleFlowArguments = [[[0, 1]], [[0, 'abc']], [[0, () => {}]]];
+
+describe('mongodb', () => {
```
Comment (doriaviram) [inline src/hooks/mongodb4x.test.js:18]: ?, follow the conventions of other test describes

File: src/hooks/mongodb4x.test.js [hunk 10]
```diff
@@ -0,0 +1,109 @@
...
+  beforeEach(() => {
+    const handlerInputs = new HandlerInputsBuilder().build();
+    TracerGlobals.setHandlerInputs(handlerInputs);
+  });
+
+  afterEach(() => {
+    connection && connection.close();
+  });
+
+  test.each(validSimpleFlowArguments)('hookMongoDb -> simple flow', async (args) => {
```
Comment (doriaviram) [inline src/hooks/mongodb4x.test.js:30]:
```suggestion
  test.each(validSimpleFlowConstractorArguments)('hookMongoDb -> simple flow', async (args) => {
```

Comment (doriaviram) [reply src/hooks/mongodb4x.test.js:30]: Don't we need to change something in the test name? I remember something that @vladlLumigo showed me.
@vladlLumigo Can u comment here?

File: testUtils/mongo4xMocker.js [hunk 11]
```diff
@@ -0,0 +1,98 @@
...
+          }
+        };
+
+        try {
+          self.originalConnect.apply(this, [self.url, {}, ...params, promiseCallbackHandler]);
+        } catch (err) {
+          reject(err);
+        }
+      });
+  }
```
Comment (doriaviram) [inline testUtils/mongo4xMocker.js:83]: Is it a total duplicate of `testUtils/mongo3xMocker.js` line 60?

Comment (doriaviram) [reply testUtils/mongo4xMocker.js:83]: There is any more duplicates with the mongo3Mocker ?

**Review by doriaviram — CHANGES_REQUESTED** [review]

File: src/extender.testModule.js [hunk 12]
```diff
@@ -0,0 +1,6 @@
+export class DummyClass {
+  value;
+  constructor() {
+    this.value = 0;
+  }
+}
```
[the remark this answers is not in the document]
Comment (therightstuff) [reply src/extender.testModule.js:6]: it won't work, it has to be in its own module

File: src/hooks/mongodb.js (hunk 3 above)
Comment (therightstuff) [reply src/hooks/mongodb.js:16]: fair enough. and i caught unexpected behavior. thanks :)

File: src/hooks/mongodb.js (hunk 4 above)
Comment (therightstuff) [reply src/hooks/mongodb.js:25]: i feel that moving it out is more readable, i think making the version detection a separate function will be cleaner and more readable so i'll try that instead

File: src/hooks/mongodb3x.test.js (hunk 7 above)
Comment (therightstuff) [reply src/hooks/mongodb3x.test.js:1]: it's a good point, but there are no behaviour changes here, just cleanup. if you'd rather i returned it let me know

File: src/hooks/mongodb3x.test.js (hunk 5 above)
Comment (therightstuff) [reply src/hooks/mongodb3x.test.js:10]: that's a fair point, reverting

File: src/hooks/mongodb3x.test.js (hunk 6 above)
Comment (therightstuff) [reply src/hooks/mongodb3x.test.js:30]: i guess i got overexcited - the 4.x client uses promises, i'll reset the 3.x client mock to what it was

File: src/hooks/mongodb4x.test.js (hunk 1 above)
Comment (therightstuff) [reply src/hooks/mongodb4x.test.js:2]: for the moment they're staying, to be discussed further

File: src/hooks/mongodb4x.test.js (hunk 10 above)
Comment (therightstuff) [reply src/hooks/mongodb4x.test.js:30]: so much better! thanks both of you 👍

File: testUtils/mongo4xMocker.js (hunk 11 above)
Comment (therightstuff) [reply testUtils/mongo4xMocker.js:83]: not any more, they're now pretty different

**Review by doriaviram — APPROVED** [review]

File: src/extender.js [hunk 13]
```diff
@@ -7,16 +7,20 @@ const noop = () => {};
 const isFunctionAlreadyWrapped = (fn) => fn && fn.__wrapped;
 
 export const hook = (module, funcName, options = {}, shimmerLib = shimmer) => {
-  const { beforeHook = noop, afterHook = noop } = options;
+  const { isConstructor = false, beforeHook = noop, afterHook = noop } = options;
```
Comment (therightstuff) [inline src/extender.js:10]: the extender can now handle constructors correctly

File: src/hooks/mongodb.test.js [hunk 14]
```diff
@@ -1,11 +1,5 @@
 import { HandlerInputsBuilder } from '../../testUtils/handlerInputsBuilder';
-import {
-  getMockedMongoClient,
-  promisifyMongoFunc,
-  wrapMongoCollection,
-} from '../../testUtils/mongoMocker';
-import { MongoSpanBuilder } from '../../testUtils/mongoSpanBuilder';
-import { SpansContainer, TracerGlobals } from '../globals';
+import { TracerGlobals } from '../globals';
```
Comment (therightstuff) [inline src/hooks/mongodb.test.js:2]: note for reviewer: client-specific tests for 3.x and 4.x clients moved to their own files

File: src/hooks/mongodb.ts [hunk 15]
```diff
@@ -0,0 +1,66 @@
+import { hook } from '../extender';
```
Comment (therightstuff) [inline src/hooks/mongodb.ts:1]: note for reviewer: renamed from mongodb.js

File: src/hooks/mongodb3x.test.js [hunk 16]
```diff
@@ -0,0 +1,97 @@
+import { HandlerInputsBuilder } from '../../testUtils/handlerInputsBuilder';
```
Comment (therightstuff) [inline src/hooks/mongodb3x.test.js:1]: note for reviewer: tests moved out from original mongodb.test.js

File: src/hooks/mongodb3x.ts [hunk 17]
```diff
@@ -0,0 +1,75 @@
+import { SpansContainer, TracerGlobals } from '../globals';
```
Comment (therightstuff) [inline src/hooks/mongodb3x.ts:1]: client-specific methods moved out from original mongodb.js

File: src/utils/payloadStringify.js [hunk 18]
```diff
@@ -43,13 +43,13 @@ export const whitelistKeysRegexes = () => {
   return keyToRegexes([], null, LUMIGO_WHITELIST_KEYS_REGEXES);
 };
 
-export const prune = (str, maxLength) => {
-  let toPrune = str;
-  if (!isString(toPrune)) {
-    logger.warn('Prune was called on a non-string object', toPrune);
-    toPrune = '';
+export const truncate = (str, maxLength, truncationString = '') => {
```
Comment (therightstuff) [inline src/utils/payloadStringify.js:46]: this is a fix for a bug discovered after a test broke intermittently when we upgraded the jest dependencies

File: src/utils/payloadStringify.js [hunk 19]
```diff
@@ -121,14 +122,14 @@ export const payloadStringify = (
...
           isPruned = true;
-          return prune(value, maxPayloadSize);
+          return truncate(value, maxPayloadSize);
         }
-        if (value instanceof Error)
+        if (value instanceof Error) {
           return {
-            stack: prune(value.stack, maxPayloadSize),
             message: value.message,
+            stack: truncate(value.stack, maxPayloadSize, TRUNCATED_TEXT),
```
Comment (therightstuff) [inline src/utils/payloadStringify.js:130]: we learned the hard way that the order of this really matters

File: testUtils/mongodbEventEmitterMocker.ts [hunk 20]
```diff
@@ -0,0 +1,12 @@
+import EventEmitter from 'events';
```
Comment (therightstuff) [inline testUtils/mongodbEventEmitterMocker.ts:1]: moved out from original mongo client mocker

**Review by nirLumigo — CHANGES_REQUESTED** [review]
Nice! :)

**Review by mmanciop — CHANGES_REQUESTED** [review]
1. Mocks in instrumentation tests are a ficus leaf that gives a false sense of security in instrumentation tests. I would straight out get rid of them and test the actual library together with testcontainers.org.
2. We need to make sure that the static `MongoClient.connect` works with the _real_ package (not the mocks), because I have seen this instrumentation break with:

```
const client = await MongoClient.connect(process.env.MONGODB_URL);
```

File: src/hooks/mongodb4x.test.js [hunk 21]
```diff
@@ -0,0 +1,114 @@
...
+      'The "url" argument must be of type string. Received type number (0)'
+    );
+    expect(SpansContainer.getSpans().length).toEqual(0);
+  });
+
+  test('hookMongoDb -> emitter "on" broken', async () => {
+    const mongoClientLibrary = getMockedMongoClientLibrary({ isOnBroken: true });
+
+    hookMongoDb(mongoClientLibrary);
+    const client = new mongoClientLibrary.MongoClient(DUMMY_URL);
```
[the remark this answers is not in the document]
Comment (therightstuff) [reply src/hooks/mongodb4x.test.js:112]: updated the test name to `broken emitter "on" method does not throw errors`

File: src/hooks/mongodb4x.ts [hunk 22]
```diff
@@ -32,26 +48,51 @@ const injectMonitoringCommand = (args: any[]) => {
...
+      // client instance
+      wrappedClass.connect = (url: string, options: any, callback: Function) => {
+        return new Promise((resolve, reject) => {
+          try {
+            callback =
+              typeof callback === 'function'
+                ? callback
+                : typeof options === 'function'
+                ? options
+                : undefined;
```
Comment (nirLumigo) [inline src/hooks/mongodb4x.ts:79]: Please don't chain ternary operators. This operator should be used only for simple cases.

File: src/hooks/mongodb4x.ts [hunk 23]
```diff
@@ -32,26 +48,51 @@ const injectMonitoringCommand = (args: any[]) => {
...
-      logger.warn(`MongoDB 4.x 'on' hooks cannot be applied to ${clientInstance}`, err);
-    }
-  } else {
-    logger.debug("MongoDB 4.x 'on' hooks skipped: monitorCommands was not set in the options");
+export const wrapMongoClient4xClass = (mongoClientLibrary: any) => {
+  try {
+    const originalStaticConnect = mongoClientLibrary.MongoClient.connect;
+    const staticProperties = getStaticProperties(mongoClientLibrary.MongoClient);
+    const wrapper = (originalFn: any) => {
+      const wrappedClass = function (...args: any[]) {
```
Comment (nirLumigo) [inline src/hooks/mongodb4x.ts:59]: If all the other functions are arrow functions, it's better to keep the inner function as arrow functions as well.

**Review by nirLumigo — APPROVED** [review]
Nice! :)

File: src/hooks/mongodb4x.ts (hunk 23 above)
Comment (doriaviram) [reply src/hooks/mongodb4x.ts:59]: I think that it can't be an arrow function. arrow function inherit the scope, and function is not

File: src/hooks/mongodb4x.ts (hunk 22 above)
Comment (doriaviram) [reply src/hooks/mongodb4x.ts:79]: Lets leave it for now
